File: lib/services/municipio_search_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/saved_location.dart';

/// Servicio para buscar municipios usando la API de AEMET.
///
/// Descarga el listado completo de municipios una única vez y lo cachea
/// en memoria para que el autocompletado sea instantáneo.
class MunicipioSearchService {
  static const String _baseUrl = 'https://opendata.aemet.es/opendata';
  static const String _apiKey =
      'eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJiaXJ0ZWJzQGdtYWlsLmNvbSIsImp0aSI6ImYwM2UxMjFmLTE2ODktNDdkMS1hYjNhLWI0MThlM2ZmMWNjMiIsImlzcyI6IkFFTUVUIiwiaWF0IjoxNzcxNDE3OTk3LCJ1c2VySWQiOiJmMDNlMTIxZi0xNjg5LTQ3ZDEtYWIzYS1iNDE4ZTNmZjFjYzIiLCJyb2xlIjoiIn0.npwJf-68OE2s0kIsRHVqjMqtmR9tedsgYrD03pjuYHc';

  final http.Client _client;

  // Caché en memoria: lista completa de municipios de AEMET.
  // Cada elemento tiene: {id, nombre, latitud_dec, longitud_dec}
  List<Map<String, dynamic>> _allMunicipios = [];
  bool _loaded = false;

  MunicipioSearchService({http.Client? client})
      : _client = client ?? http.Client();

  /// Carga el listado completo si aún no está en caché.
  Future<void> _ensureLoaded() async {
    if (_loaded) return;

    const endpoint = '$_baseUrl/api/maestro/municipios';

    final response1 = await _client.get(
      Uri.parse(endpoint),
      headers: {'api_key': _apiKey},
    );

    if (response1.statusCode != 200) return;

    final body1 = jsonDecode(response1.body) as Map<String, dynamic>;
    final datosUrl = body1['datos'] as String?;
    if (datosUrl == null) return;

    final response2 = await _client.get(Uri.parse(datosUrl));
    if (response2.statusCode != 200) return;

    String decoded;
    try {
      decoded = utf8.decode(response2.bodyBytes);
    } catch (_) {
      decoded = latin1.decode(response2.bodyBytes);
    }

    final data = jsonDecode(decoded);
    if (data is List) {
      _allMunicipios = data.cast<Map<String, dynamic>>();
    }
    _loaded = true;
  }

  /// Busca municipios cuyo nombre contenga [query] (insensible a mayúsculas).
  ///
  /// Retorna hasta 10 resultados.
  Future<List<SavedLocation>> searchByName(String query) async {
    if (query.trim().isEmpty) return [];

    await _ensureLoaded();

    final q = _normalize(query);

    return _allMunicipios
        .where((m) {
          final nombre = _normalize(m['nombre'] as String? ?? '');
          return nombre.contains(q);
        })
        .take(10)
        .map((m) => SavedLocation(
              municipioId: (m['id'] as String).replaceFirst('id', ''),
              nombre: _cleanNombre(m['nombre'] as String? ?? ''),
            ))
        .toList();
  }

  /// Encuentra el municipio más cercano a las coordenadas dadas.
  Future<SavedLocation?> findNearestMunicipio(double lat, double lon) async {
    await _ensureLoaded();
    if (_allMunicipios.isEmpty) return null;

    SavedLocation? nearest;
    double minDist = double.infinity;

    for (final m in _allMunicipios) {
      final latStr = m['latitud_dec'] as String?;
      final lonStr = m['longitud_dec'] as String?;
      if (latStr == null || lonStr == null) continue;

      final mLat = double.tryParse(latStr.replaceAll(',', '.'));
      final mLon = double.tryParse(lonStr.replaceAll(',', '.'));
      if (mLat == null || mLon == null) continue;

      final dist = _dist(lat, lon, mLat, mLon);
      if (dist < minDist) {
        minDist = dist;
        nearest = SavedLocation(
          municipioId: (m['id'] as String).replaceFirst('id', ''),
          nombre: _cleanNombre(m['nombre'] as String? ?? ''),
        );
      }
    }

    return nearest;
  }

  /// Distancia euclídea sobre grados (suficiente para este uso).
  double _dist(double lat1, double lon1, double lat2, double lon2) {
    final dlat = lat1 - lat2;
    final dlon = lon1 - lon2;
    return dlat * dlat + dlon * dlon;
  }

  /// Normaliza el texto para comparación: minúsculas, sin tildes.
  String _normalize(String s) {
    return s
        .toLowerCase()
        .replaceAll('á', 'a')
        .replaceAll('é', 'e')
        .replaceAll('í', 'i')
        .replaceAll('ó', 'o')
        .replaceAll('ú', 'u')
        .replaceAll('ü', 'u')
        .replaceAll('à', 'a')
        .replaceAll('è', 'e')
        .replaceAll('ì', 'i')
        .replaceAll('ò', 'o')
        .replaceAll('ù', 'u')
        .replaceAll('ñ', 'n');
  }

  /// Limpia el nombre AEMET: elimina el prefijo de provincia entre paréntesis.
  /// Ej: "Madrid" queda igual; "Valencia/València" se queda como está.
  String _cleanNombre(String nombre) {
    // AEMET a veces incluye "(provincia)" al final
    return nombre.replaceAll(RegExp(r'\s*\(.*?\)\s*$'), '').trim();
  }
  /// Obtiene las coordenadas (latitud, longitud) de un municipio por su ID.
  Future<({double lat, double lon})?> getCoordinates(String municipioId) async {
    await _ensureLoaded();
    
    // El id en el JSON tiene prefijo "id" (ej: "id28079")
    final searchId = municipioId.startsWith('id') ? municipioId : 'id$municipioId';
    
    try {
      final municipio = _allMunicipios.firstWhere(
        (m) => m['id'] == searchId,
        orElse: () => {},
      );
      
      if (municipio.isEmpty) return null;
      
      final latStr = municipio['latitud_dec'] as String?;
      final lonStr = municipio['longitud_dec'] as String?;
      
      if (latStr == null || lonStr == null) return null;
      
      final lat = double.tryParse(latStr.replaceAll(',', '.'));
      final lon = double.tryParse(lonStr.replaceAll(',', '.'));
      
      if (lat != null && lon != null) {
        return (lat: lat, lon: lon);
      }
    } catch (_) {
      return null;
    }
    return null;
  }
}

File: app/src/main/java/com/nubo/nubo/data/remote/MunicipioSearchService.kt
package com.nubo.nubo.data.remote

import com.nubo.nubo.domain.model.SavedLocation
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray
import org.json.JSONObject

/** Coordenadas de un municipio. */
data class Coordinates(val lat: Double, val lon: Double)

/**
 * Búsqueda de municipios españoles contra el maestro de AEMET.
 *
 * El listado completo (unos 8.000 municipios) se descarga una sola vez y se
 * mantiene en memoria, de modo que el autocompletado no toque la red en cada
 * pulsación.
 */
class MunicipioSearchService(
    private val aemet: AemetApi = AemetApi(),
) {
    private var municipios: List<Municipio> = emptyList()
    private var loaded = false

    // Evita que varias búsquedas simultáneas lancen varias descargas.
    private val loadMutex = Mutex()

    /** Una entrada del maestro, ya normalizada. */
    internal data class Municipio(
        val id: String,
        val nombre: String,
        /**
         * Formas contra las que se busca.
         *
         * Son dos porque AEMET pospone el artículo ("Palmas de Gran Canaria,
         * Las") pero la app muestra la forma natural ("Las Palmas de Gran
         * Canaria"). Indexando solo la cruda, buscar "Las Palmas" no encontraba
         * nada; indexando solo la mostrada, fallaría quien escriba "Palmas,".
         */
        val searchable: List<String>,
        val lat: Double?,
        val lon: Double?,
        /**
         * Zona de aviso de AEMET a la que pertenece el municipio.
         *
         * Seis dígitos: área (2) + provincia (2) + zona dentro de la provincia
         * (2). Es el mismo código que viaja en el `geocode` de cada aviso CAP,
         * así que es lo que permite quedarse solo con los avisos que de verdad
         * afectan al municipio. Ver [AlertService].
         */
        val zonaAviso: String?,
    )

    private suspend fun ensureLoaded() {
        if (loaded) return
        loadMutex.withLock {
            if (loaded) return

            val body = try {
                aemet.fetchData("/api/maestro/municipios", timeoutSeconds = 30)
            } catch (_: Exception) {
                null
            } ?: return

            municipios = try {
                parseMaestro(JSONArray(body))
            } catch (_: Exception) {
                return
            }
            loaded = true
        }
    }

    internal fun parseMaestro(array: JSONArray): List<Municipio> {
        val result = ArrayList<Municipio>(array.length())
        for (i in 0 until array.length()) {
            val item = array.optJSONObject(i) ?: continue
            val rawId = item.optString("id").takeIf { it.isNotBlank() } ?: continue
            val nombre = cleanNombre(item.optString("nombre"))
            if (nombre.isBlank()) continue

            val display = SavedLocation.formatNombre(nombre)
            result += Municipio(
                // El maestro trae el id con prefijo, p. ej. "id28079".
                id = rawId.removePrefix("id"),
                nombre = nombre,
                searchable = listOf(normalize(nombre), normalize(display)).distinct(),
                lat = parseDecimal(item, "latitud_dec"),
                lon = parseDecimal(item, "longitud_dec"),
                zonaAviso = parseZonaAviso(item),
            )
        }
        return result
    }

    /** Hasta 10 municipios cuyo nombre contenga [query], sin distinguir tildes. */
    suspend fun searchByName(query: String): List<SavedLocation> {
        if (query.isBlank()) return emptyList()
        ensureLoaded()

        val q = normalize(query)
        return municipios
            .asSequence()
            .filter { municipio -> municipio.searchable.any { it.contains(q) } }
            .take(MAX_RESULTS)
            .map { SavedLocation.of(it.id, it.nombre) }
            .toList()
    }

    /** Municipio más próximo a unas coordenadas. */
    suspend fun findNearestMunicipio(lat: Double, lon: Double): SavedLocation? {
        ensureLoaded()

        var nearest: Municipio? = null
        var minDistance = Double.MAX_VALUE

        for (municipio in municipios) {
            val mLat = municipio.lat ?: continue
            val mLon = municipio.lon ?: continue

            // Distancia euclídea en grados: no es geodésica, pero para comparar
            // municipios dentro de España el orden que produce es el mismo.
            val dLat = lat - mLat
            val dLon = lon - mLon
            val distance = dLat * dLat + dLon * dLon

            if (distance < minDistance) {
                minDistance = distance
                nearest = municipio
            }
        }

        return nearest?.let { SavedLocation.of(it.id, it.nombre) }
    }

    /**
     * Zona de aviso de AEMET de un municipio, o `null` si el maestro no la
     * trae o no se pudo descargar.
     */
    suspend fun getZonaAviso(municipioId: String): String? {
        if (municipioId.isBlank()) return null
        ensureLoaded()

        val id = municipioId.removePrefix("id")
        return municipios.firstOrNull { it.id == id }?.zonaAviso
    }

    /** Coordenadas de un municipio por su código INE. */
    suspend fun getCoordinates(municipioId: String): Coordinates? {
        if (municipioId.isBlank()) return null
        ensureLoaded()

        val id = municipioId.removePrefix("id")
        val municipio = municipios.firstOrNull { it.id == id } ?: return null
        val lat = municipio.lat ?: return null
        val lon = municipio.lon ?: return null
        return Coordinates(lat, lon)
    }

    internal companion object {
        const val MAX_RESULTS = 10

        /**
         * Lee `zona_comarcal` del maestro, exigiendo los 6 dígitos.
         *
         * Se valida el formato en vez de aceptar lo que venga porque de este
         * campo depende qué avisos se muestran: un valor raro debe dejar al
         * municipio sin avisos, nunca colarle los de otra zona.
         */
        fun parseZonaAviso(item: JSONObject): String? =
            item.optString("zona_comarcal").trim().takeIf { ZONA_AVISO.matches(it) }

        /** Área (2) + provincia (2) + zona (2). */
        val ZONA_AVISO = Regex("""\d{6}""")

        /** AEMET publica los decimales con coma en algunos registros. */
        fun parseDecimal(item: JSONObject, key: String): Double? =
            item.optString(key).takeIf { it.isNotBlank() }
                ?.replace(',', '.')
                ?.toDoubleOrNull()

        /** Quita el paréntesis final con la provincia, si lo trae. */
        fun cleanNombre(nombre: String): String =
            nombre.replace(Regex("""\s*\(.*?\)\s*$"""), "").trim()

        private val ACCENTS = mapOf(
            'á' to 'a', 'à' to 'a', 'ä' to 'a', 'â' to 'a',
            'é' to 'e', 'è' to 'e', 'ë' to 'e', 'ê' to 'e',
            'í' to 'i', 'ì' to 'i', 'ï' to 'i', 'î' to 'i',
            'ó' to 'o', 'ò' to 'o', 'ö' to 'o', 'ô' to 'o',
            'ú' to 'u', 'ù' to 'u', 'ü' to 'u', 'û' to 'u',
            'ñ' to 'n', 'ç' to 'c',
        )

        /** Minúsculas y sin tildes, para que "Coruna" encuentre "A Coruña". */
        fun normalize(s: String): String =
            s.lowercase().map { ACCENTS[it] ?: it }.joinToString("")
    }
}
